lib/tpm_policy.c

```diff
@@ -0,0 +1,269 @@
+            pctx->common_policy_options.policy_session_type, &symmetric, 
+            pctx->common_policy_options.policy_digest_hash_alg);
+    if (rval != TPM_RC_SUCCESS) {
+        LOG_ERR("Failed tpm session start auth with params\n");
+    }
+    return rval;
+}
+
+TPM_RC build_policy(CREATE_POLICY_CTX *pctx,
+        TPM_RC (*build_policy_function)(CREATE_POLICY_CTX *pctx)) {    
+    //Start policy session
+    TPM_RC rval = start_policy_session(pctx);
+    if (rval != TPM_RC_SUCCESS) {
+        LOG_ERR("Error starting the policy session.\n");
+        goto build_policy_error;
+    }
+    // Issue policy command.
+    rval = (*build_policy_function)(pctx);
+    if (rval != TPM_RC_SUCCESS) {
+        LOG_ERR("Failed parse_policy_type_and_send_command\n");
+        goto build_policy_error;
+    }
+    // Get Policy Hash
+    rval = Tss2_Sys_PolicyGetDigest(pctx->sapi_context,
+            pctx->common_policy_options.policy_session->sessionHandle, 0, 
+            &pctx->common_policy_options.policy_digest, 0);
+    if (rval != TPM_RC_SUCCESS) {
+        LOG_ERR("Failed Policy Get Digest\n");
+        goto build_policy_error;
+    } 
+    // Display the policy digest during real policy session.
+    if (pctx->common_policy_options.policy_session_type == TPM_SE_POLICY) {
+        printf("TPM_SE_POLICY: 0x");
+        int i;
+        for(i=0; i<pctx->common_policy_options.policy_digest.t.size; i++) {
+            printf("%02X", pctx->common_policy_options.policy_digest.t.buffer[i]);
+        }
+        printf("\n");
+    }
+    // Additional operations when session if a trial policy session
+    if (pctx->common_policy_options.policy_session_type == TPM_SE_TRIAL) {
+        //save the policy buffer in a file for use later
+        bool result = files_save_bytes_to_file(pctx->common_policy_options.policy_file, 
+            (UINT8 *) &pctx->common_policy_options.policy_digest.t.buffer, 
+            pctx->common_policy_options.policy_digest.t.size);
+        if (!result) {
+            LOG_ERR("Failed to save policy digest into file \"%s\"",
+                    pctx->common_policy_options.policy_file);
+            rval = TPM_RC_NO_RESULT;
+            goto build_policy_error;
+        }
+    }
+    // Need to flush the session here.
+    if (!pctx->common_policy_options.extend_policy_session) {
+        rval = Tss2_Sys_FlushContext(pctx->sapi_context,
+                pctx->common_policy_options.policy_session->sessionHandle);
+        if (rval != TPM_RC_SUCCESS) {
+            LOG_ERR("Failed Flush Context\n");
+            goto build_policy_error;
+        }             
+    } 
+    // And remove the session from sessions table.
+    rval = tpm_session_auth_end(pctx->common_policy_options.policy_session);
+    if (rval != TPM_RC_SUCCESS) {
+        LOG_ERR("Failed deleting session from session table\n");        
+        goto build_policy_error;
+    }
+
+build_policy_error:
+    return rval;
```

Comment:
The goto doesn't do anything but return, so just replace your goto with return rval;